// mui-tiptap/dist/cjs/extensions/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableImproved = exports.ResizableImage = exports.LinkBubbleMenuHandler = exports.scrollToCurrentHeadingAnchor = exports.HeadingWithAnchor = exports.FontSize = void 0;
var FontSize_1 = require("./FontSize");
Object.defineProperty(exports, "FontSize", { enumerable: true, get: function () { return __importDefault(FontSize_1).default; } });
var HeadingWithAnchor_1 = require("./HeadingWithAnchor");
Object.defineProperty(exports, "HeadingWithAnchor", { enumerable: true, get: function () { return __importDefault(HeadingWithAnchor_1).default; } });
Object.defineProperty(exports, "scrollToCurrentHeadingAnchor", { enumerable: true, get: function () { return HeadingWithAnchor_1.scrollToCurrentHeadingAnchor; } });
var LinkBubbleMenuHandler_1 = require("./LinkBubbleMenuHandler");
Object.defineProperty(exports, "LinkBubbleMenuHandler", { enumerable: true, get: function () { return __importDefault(LinkBubbleMenuHandler_1).default; } });
var ResizableImage_1 = require("./ResizableImage");
Object.defineProperty(exports, "ResizableImage", { enumerable: true, get: function () { return __importDefault(ResizableImage_1).default; } });
var TableImproved_1 = require("./TableImproved");
Object.defineProperty(exports, "TableImproved", { enumerable: true, get: function () { return __importDefault(TableImproved_1).default; } });
