// mui-tiptap/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getImageBackgroundColorStyles = exports.getEditorStyles = exports.Z_INDEXES = exports.useRichTextEditorContext = exports.RichTextEditorContext = exports.TableBubbleMenu = exports.RichTextReadOnly = exports.RichTextField = exports.RichTextEditorProvider = exports.RichTextEditor = exports.RichTextContent = exports.MenuDivider = exports.MenuBar = exports.LinkBubbleMenu = exports.ControlledBubbleMenu = void 0;
var ControlledBubbleMenu_1 = require("./ControlledBubbleMenu");
Object.defineProperty(exports, "ControlledBubbleMenu", { enumerable: true, get: function () { return __importDefault(ControlledBubbleMenu_1).default; } });
var LinkBubbleMenu_1 = require("./LinkBubbleMenu");
Object.defineProperty(exports, "LinkBubbleMenu", { enumerable: true, get: function () { return __importDefault(LinkBubbleMenu_1).default; } });
var MenuBar_1 = require("./MenuBar");
Object.defineProperty(exports, "MenuBar", { enumerable: true, get: function () { return __importDefault(MenuBar_1).default; } });
var MenuDivider_1 = require("./MenuDivider");
Object.defineProperty(exports, "MenuDivider", { enumerable: true, get: function () { return __importDefault(MenuDivider_1).default; } });
var RichTextContent_1 = require("./RichTextContent");
Object.defineProperty(exports, "RichTextContent", { enumerable: true, get: function () { return __importDefault(RichTextContent_1).default; } });
var RichTextEditor_1 = require("./RichTextEditor");
Object.defineProperty(exports, "RichTextEditor", { enumerable: true, get: function () { return __importDefault(RichTextEditor_1).default; } });
var RichTextEditorProvider_1 = require("./RichTextEditorProvider");
Object.defineProperty(exports, "RichTextEditorProvider", { enumerable: true, get: function () { return __importDefault(RichTextEditorProvider_1).default; } });
var RichTextField_1 = require("./RichTextField");
Object.defineProperty(exports, "RichTextField", { enumerable: true, get: function () { return __importDefault(RichTextField_1).default; } });
var RichTextReadOnly_1 = require("./RichTextReadOnly");
Object.defineProperty(exports, "RichTextReadOnly", { enumerable: true, get: function () { return __importDefault(RichTextReadOnly_1).default; } });
var TableBubbleMenu_1 = require("./TableBubbleMenu");
Object.defineProperty(exports, "TableBubbleMenu", { enumerable: true, get: function () { return __importDefault(TableBubbleMenu_1).default; } });
var context_1 = require("./context");
Object.defineProperty(exports, "RichTextEditorContext", { enumerable: true, get: function () { return context_1.RichTextEditorContext; } });
Object.defineProperty(exports, "useRichTextEditorContext", { enumerable: true, get: function () { return context_1.useRichTextEditorContext; } });
__exportStar(require("./controls"), exports);
__exportStar(require("./extensions"), exports);
__exportStar(require("./hooks"), exports);
var styles_1 = require("./styles");
Object.defineProperty(exports, "Z_INDEXES", { enumerable: true, get: function () { return styles_1.Z_INDEXES; } });
Object.defineProperty(exports, "getEditorStyles", { enumerable: true, get: function () { return styles_1.getEditorStyles; } });
Object.defineProperty(exports, "getImageBackgroundColorStyles", { enumerable: true, get: function () { return styles_1.getImageBackgroundColorStyles; } });
__exportStar(require("./utils"), exports);
